### DISCUSSION TITLE
What are the advantages of switching from Plotly Dash to idom?

### DISCUSSION THREAD

**Discussion by pikhovkin** [idom-team/idom#473] [Question] [answered]
What are the advantages of switching from Plotly Dash to idom?

**Comment by rmorshea** [comment, marked as answer]
So first off, you can actually work with IDOM in a Dash application using [`idom-dash`](https://github.com/idom-team/idom-dash) 🙂

To answer your question though, why would someone want to use IDOM with or instead of Dash? Well, I'm not especially familiar with Dash, but from what I understand...

| In IDOM                                                                                                                                                                                                                                        | In Dash                                                                                                                                                                                                                                                                  |
|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| You can write fully functional components purely in Python                                                                                                                                                                                     | To create a completely new dash component, you have to write Javascript. Further, they're [tooling](https://github.com/plotly/dash-component-boilerplate) for doing this isn't especially user friendly.                                                                 |
| You can easily use components developed with IDOM on several different Platforms, in standalone apps, or embedded in existing apps.                                                                                                              | Being able to use your Dash components elsewhere is pretty difficult. For example, what if you wanted to use them in a Jupyter Notebook or embed them into some non-Dash-based app.                                                                                      |
| The server can trigger updates to the client-side view (i.e. the Server can be in the driver's seat). For example, you can create [animated dashboards](https://idom-docs.herokuapp.com/docs/examples.html#simple-dashboard) purely in Python. | The server cannot update the view without the client sending it a signal first. This is because Dash doesn't have true bidirectional communication (like one can get from a WebSocket).                                                                                  |
| You can manage and share state between components in IDOM using the same tried-and-true patterns as React                                                                                                                                      | Dealing with state and dependencies between components is clunky, leading to questions like [this](https://community.plotly.com/t/how-to-elegantly-handle-a-very-large-number-of-input-state-in-callbacks/19228) where none of the solutions really look all that clean. |

If you'd like to learn more I've written about some of the [motivations for creating IDOM](https://idom-docs.herokuapp.com/docs/architectural-patterns.html).

**Comment by kamilc** [comment]
@rmorshea Is there any way to use Dash callbacks with `idom-dash` components?

**Reply by rmorshea** [reply]
I'm not sure the stateless design philosophy of Dash callbacks fits very well with IDOM. For example, what would you do with the output of the callback? At the moment, it seems like [Dash does not support callbacks that lack an output](https://community.plotly.com/t/app-callback-without-an-output/5502/23).

With that said, in principal the following would work:

```python
class DashState():

    def __init__(self, output: dash.Output, *inputs: dash.Input) -> None:
        @dash.callback(output, *inputs)
        def handle_input_changes(*args):
            for loop, queue in self._subscribers:
                loop.call_soon_threadsafe(
                    queue.put_nowait,
                    args[0] if len(args) == 1 else args
                )

        self._num_inputs = len(inputs)
        self._subscribers = []

    def use_state(self) -> Any:
        state, set_state = use_state((None,) * self._num_inputs)

        @use_effect
        async def add_queue():
            queue = Queue()
            self._subscribers.append((asyncio.get_running_loop(), queue))
            while True:
                set_state(await queue.get())

        return state
```

Before adding this into `idom_dash`, it would be useful to know what your use case is to see if this would really solve it.
